### PR TITLE
MAINT: Refactor Plugin Loading

### LINKED ISSUE #307: Programmatic Plugin loading
Currently the way the plugins are loaded is very restrictive. You have to have files named with `_plugin.py`, no where these files are definitively in the file structure AND set your environment variable to look for these before you IPython session even starts. 

I'm proposing something a little more flexible such that you could:
```python
from pydm.data_plugins import add_plugin
add_plugin(MyCustomPlugin)
```
This allows a more programatic way to add plugins with out screwing around with the file system and environment variables. The other change that this would require is that currently the PyDMApplication loads these plugins at the dawn of time and then doesn't update. This leads to an order of operations problem.

```python
from pydm import PyDMApplication  # Loads all plugins
pydm.data_plugins.plugins.append(MyCustomPlugin)  # Will not be added to plugin registry
```
This would have have to be rectified as well so that plugins added after import are found.

### PR DESCRIPTION
**Pull request by teddyrendahl**
## Description
Refactored the way plugins are loaded. There are now two methods:

* ``add_plugin``
     - Add a plugin to the global registry
* ``load_plugins_from_path``
     -  Searches a list of directories for valid `PyDMPlugin`objects. This is automatically called in `pydm/data_plugins` as well as any directories listed in`$PYDM_DATA_PLUGIN_PATH`.

Finally, the last major change is now instead of constructing the mapping of protocol vs. plugins in the `PyDMApplication` itself, `PyDMApplication.plugins` is just linked to `data_plugins.plugin_modules`. `plugin_modules` was formerly a `list` is now a `dict`. This allows plugins to be added after `PyDMApplication` is imported. This is explicitly tested now in `test_add_plugin`.

### Question
The way this was loading plugins before enforced a few restrictions:
1. Can not overwrite existing protocols
2. Can not have multiple plugins come from a single.

I find both of these to be kind of arbitrary. The first seems like it could be reasonable to want i.e different facilities have different `arch` data plugins. As long as there is a proper warning generated it seems like this should be tolerated. Second, just seems like a strange thing to check for and deleting that might clean the code up.

## Motivation
Closes #307 

## Documentation
I couldn't find a page about how these plugins are loaded so added one myself. Please let me know if this is in the correct location.

## Tests
Created `test_data_plugins_import`. These are slightly unnecessary because if plugin loading fails we'll see a cascade of errors in other tests but still fine for completeness.

### CODE REVIEW

**Comment by hhslepicka** [comment]
Great work, @teddyrendahl!

About the questions...
1-As we discussed offline, if a proper warning is generated it is good enough for me. Users should be responsible for ensuring proper usage of plugins.
2-Before we were all up a 1-1 relation between a plugin and a file but for me it is okay to accept more than one on a file. Can you take care of this one and remove the limitation?

**Comment by teddyrendahl** [comment]
Sounds good. I'll modify this and add a few more tests for the edge cases as well.

**Comment by teddyrendahl** [comment]
Alright. Flushed out the tests and now we allow more than one `PyDMPlugin` per file. We also warn the user when they overwrite an existing protocol but we still allow them to do so.

File: pydm/tests/test_data_plugins_import.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    # Create test PyDMPlugin with mock protocol
+    test_plug = PyDMPlugin
+    test_plug.protocol = 'tst'
+    # Check that adding this after import will be reflected in PyDMApp
+    add_plugin(test_plug)
+    assert isinstance(plugin_modules['tst'], test_plug)
+    assert isinstance(qapp.plugins['tst'], test_plug)
+
+
+def test_plugin_directroy_loading(qapp):
```
Comment (hhslepicka) [inline pydm/tests/test_data_plugins_import.py]: Typo in test name: directroy -> directory

**Comment by teddyrendahl** [comment]
Issue has been addressed with Windows. Instead of wasting time with `tempfile` strangeness just made a regular file which seems to work. Ready for final review.

**Comment by hhslepicka** [comment]
@mattgibbs and I talked offline today and we both agree with it so I will go ahead and merge the PR.
Thank you for the great work, @teddyrendahl .
